### PR TITLE
feature/CP-4533-version-comparison-code

### PR DESCRIPTION
**Pull request by cpandya25**
CP-4533 App version comparator class added with comparison.

### CODE REVIEW

File: CleverPush/CleverPush/Source/CPAppBannerModuleInstance.m [hunk 1]
```diff
@@ -290,31 +290,28 @@ - (BOOL)checkRelationFilter:(NSString*)value compareWith:(NSString*)compareValue
...
             allowed = NO;
         }
     } else if (allowed && [relation isEqualToString:filterRelationType(CPFilterRelationTypeBetween)]) {
-        if (resultFrom != NSOrderedSame && resultFrom != NSOrderedDescending && resultTo != NSOrderedSame && resultTo != NSOrderedDescending) {
+        if ([value isEqualToVersion:compareValueFrom] == NO && [value isEqualOrOlderThanVersion:compareValueFrom] == TRUE && [value isEqualToVersion:compareValueTo] == NO && [value isEqualOrOlderThanVersion:compareValueTo] == TRUE) {
             allowed = NO;
         }
     } else if (allowed && [relation isEqualToString:filterRelationType(CPFilterRelationTypeNotEqual)]) {
-        if (allowed && result == NSOrderedSame) {
+        if (allowed && [value isEqualToVersion:compareValue] == TRUE) {
```
Comment (gbhrdt) [inline CleverPush/CleverPush/Source/CPAppBannerModuleInstance.m:314]: YES

File: CleverPush/CleverPush/Source/CPAppBannerModuleInstance.m [hunk 2]
```diff
@@ -290,31 +290,28 @@ - (BOOL)checkRelationFilter:(NSString*)value compareWith:(NSString*)compareValue
...
         return allowed;
     }
     if (allowed && [relation isEqualToString:filterRelationType(CPFilterRelationTypeEquals)]) {
-        if (allowed && result != NSOrderedSame) {
+        if (allowed && [value isEqualToVersion:compareValue] == NO) {
             allowed = NO;
         }
     } else if (allowed && [relation isEqualToString:filterRelationType(CPFilterRelationTypeGreaterThan)]) {
-        if (result != NSOrderedDescending) {
+        if ([value isEqualOrOlderThanVersion:compareValue] == TRUE) {
```
Comment (gbhrdt) [inline CleverPush/CleverPush/Source/CPAppBannerModuleInstance.m:302]: YES

File: CleverPush/CleverPush/Source/CPAppBannerModuleInstance.m [hunk 3]
```diff
@@ -290,31 +290,28 @@ - (BOOL)checkRelationFilter:(NSString*)value compareWith:(NSString*)compareValue
...
             allowed = NO;
         }
     } else if (allowed && [relation isEqualToString:filterRelationType(CPFilterRelationTypeLessThan)]) {
-        if (result != NSOrderedAscending) {
+        if ([value isEqualOrNewerThanVersion:compareValue] == TRUE) {
```
Comment (gbhrdt) [inline CleverPush/CleverPush/Source/CPAppBannerModuleInstance.m:306]: YES

**Review by gbhrdt — CHANGES_REQUESTED** [review]

File: CleverPush/Source/CPAppVersionComparator.m [hunk 4]
```diff
@@ -0,0 +1,49 @@
+#import "CPAppVersionComparator.h"
+
+static NSString *versionSeparator = @".";
+
+@implementation NSString (CompareToVersion)
+
+- (NSComparisonResult) CompareToVersion:(NSString *)version {
```
Comment (gbhrdt) [inline CleverPush/Source/CPAppVersionComparator.m:7]: `- (NSComparisonResult)compareToVersion:(NSString *)version {`

File: CleverPush/Source/CPAppVersionComparator.h [hunk 5]
```diff
@@ -0,0 +1,13 @@
+#import <Foundation/Foundation.h>
+
+@interface NSString (CompareToVersion)
+
+- (NSComparisonResult)CompareToVersion:(NSString *)version;
```
Comment (gbhrdt) [inline CleverPush/Source/CPAppVersionComparator.h:5]: `- (NSComparisonResult)compareToVersion:(NSString *)version {`

File: CleverPush/Source/CPAppBannerModuleInstance.m [hunk 6]
```diff
@@ -2,7 +2,7 @@
 #import "CPUtils.h"
 #import "CPLog.h"
 #import "NSDictionary+SafeExpectations.h"
-
+#import "CPAppVersionComparator.h"
```
Comment (gbhrdt) [inline CleverPush/Source/CPAppBannerModuleInstance.m:5]: Add new line after

File: CleverPush/Source/CPAppBannerModuleInstance.m [hunk 7]
```diff
@@ -279,38 +279,39 @@ - (BOOL)bannerTargetingAllowed:(CPAppBanner*)banner {
...
             allowed = NO;
         }
     } else if (allowed && [relation isEqualToString:filterRelationType(CPFilterRelationTypeGreaterThan)]) {
-        if (!CHECK_FILTER_GREATER_THAN(value, compareValue)) {
+        if ([value isEqualOrOlderThanVersion:compareValue] == YES) {
             allowed = NO;
         }
     } else if (allowed && [relation isEqualToString:filterRelationType(CPFilterRelationTypeLessThan)]) {
-        if (!CHECK_FILTER_LESS_THAN(value, compareValue)) {
+        if ([value isEqualOrNewerThanVersion:compareValue] == YES) {
```
Comment (gbhrdt) [inline CleverPush/Source/CPAppBannerModuleInstance.m:306]: Actually we don't need `== YES` We can just do
`if ([value isEqualOrNewerThanVersion:compareValue]) {`

Also apply to all other conditions please.

Same with NO, just check if `!condition`

**Review by gbhrdt — CHANGES_REQUESTED** [review]

File: CleverPush/Source/CPAppVersionComparator.m [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+        }
+        if (thisSegment > compareSegment) {
+            return NSOrderedDescending;
+        }
+    }
+    return NSOrderedSame;
+}
+
+- (BOOL)isOlderThanVersion:(NSString *)version {
+    return ([self compareToVersion:version] == NSOrderedAscending);
```
Comment (cpKevinrodriguesnobre) [inline CleverPush/Source/CPAppVersionComparator.m:30]: () not necessary here and below

File: CleverPush/Source/CPAppVersionComparator.m [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+- (BOOL)isNewerThanVersion:(NSString *)version {
+    return ([self compareToVersion:version] == NSOrderedDescending);
```
Comment (cpKevinrodriguesnobre) [inline CleverPush/Source/CPAppVersionComparator.m:34]: ()

File: CleverPush/Source/CPAppVersionComparator.m [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+- (BOOL)isEqualToVersion:(NSString *)version {
+    return ([self compareToVersion:version] == NSOrderedSame);
```
Comment (cpKevinrodriguesnobre) [inline CleverPush/Source/CPAppVersionComparator.m:38]: ()

File: CleverPush/Source/CPAppVersionComparator.m [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+- (BOOL)isEqualOrOlderThanVersion:(NSString *)version {
+    return ([self compareToVersion:version] != NSOrderedDescending);
```
Comment (cpKevinrodriguesnobre) [inline CleverPush/Source/CPAppVersionComparator.m:42]: ()

File: CleverPush/Source/CPAppVersionComparator.m [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+- (BOOL)isEqualOrNewerThanVersion:(NSString *)version {
+    return ([self compareToVersion:version] != NSOrderedAscending);
```
Comment (cpKevinrodriguesnobre) [inline CleverPush/Source/CPAppVersionComparator.m:46]: ()

File: CleverPush/Source/CPAppVersionComparator.m [hunk 13]
```diff
@@ -0,0 +1,49 @@
+#import "CPAppVersionComparator.h"
+
+static NSString *versionSeparator = @".";
+
+@implementation NSString (compareToVersion)
+
+- (NSComparisonResult)compareToVersion:(NSString *)version {
+    if ([self isEqualToString:version])
```
Comment (cpKevinrodriguesnobre) [inline CleverPush/Source/CPAppVersionComparator.m:8]: always use {} here
if(){
}

**Review by cpKevinrodriguesnobre — CHANGES_REQUESTED** [review]

**Review by cpKevinrodriguesnobre — APPROVED** [review]

**Pull request merged**